oligo_designer_toolsuite/oligo_specificity_filter/_filter_base.py

```diff
@@ -0,0 +1,170 @@
+from Bio.Seq import Seq
+from Bio.SeqRecord import SeqRecord
+
+from ..IO._data_parser import write_oligos_DB_tsv
+
+
+class ProbeFilterBase(ABC):
+    def __init__(
+        self,
+        n_jobs,
+        dir_output,
+        dir_annotations=None,
+        number_subbatches=None,
+        max_genes_in_batch=300,
+    ):
+        """This is the base class for all probe filter classes
+
+        :param n_jobs: numbers of jobs to run in parallel when filtering the probes
+        :type n_jobs: int
+        :param dir_output: output directory to write file containing filtered probes to
+        :type dir_output: str
+        :param dir_annotations: directory storing file annotations of probes, defaults to None
+        :type dir_annotations: str, optional
+        :param number_subbatches: number of subbatches to split batched data into to control how many genes are in one batch, defaults to None
+        :type number_subbatches: int, optional
+        :param max_genes_in_batch: max number of genes allowed in a batch, defaults to 300
+        :type max_genes_in_batch: int, optional
+        """
+        self.dir_output = dir_output
+        self.max_genes_in_batch = max_genes_in_batch  # if more than 300 genes in one batch split into subbatches to reduce required memory for loading blast results
+        self.duplicated_sequences = None
+        self.number_subbatches = number_subbatches
+
+        if n_jobs == None:
+            self.n_jobs = joblib.cpu_count()
+        else:
+            self.n_jobs = n_jobs
+
+        # set logger
+        log_file = os.path.join(self.dir_output, "logs/filter_probes.log")
+        file_handler = logging.FileHandler(filename=log_file)
+        stdout_handler = logging.StreamHandler(stream=sys.stdout)
+        handlers = [file_handler, stdout_handler]
+
+        logging.basicConfig(level=logging.INFO, handlers=handlers)
+        self.logging = logging.getLogger("filter_probes")
+
+        # set directory
+        if dir_annotations == None:
+            self.dir_annotations = os.path.join(self.dir_output, "annotations")
+            Path(self.dir_annotations).mkdir(parents=True, exist_ok=True)
+        else:
+            self.dir_annotations = dir_annotations
+
+    @abstractmethod
+    def apply(self, probe_info):
+        """Apply filter to list of all possible probes in probe_info dictionary given user-specified genes and save output in dir_output
+
+        :param probe_info: probe info of user-specified genes
+        :type probe_info: dict
+        """
+
+    def create_batches(self, probe_info):
+        """Create batches of files that subdivide the probes and probe info such that all probes of a gene are in one batch. These batches are then used as input to the filters. Number of batches is equal to n_jobs.
+
+        :param probe_info: probe info of user-specified genes
+        :type probe_info: dict
+        """
+
+        write_oligos_DB_tsv(probe_info, self.dir_annotations + "/tmp.tsv")
```

Comment:
this part I don't get. Why do we write, read and then delete? Instead we should use the dictionary database, no?